src/services/pcn-router/datamodel/router.yang

```diff
@@ -23,11 +23,20 @@ module router {
         type inet:ipv4-prefix;
         description "IP address and prefix of the port";
         polycube-base:cli-example "10.0.0.1/24";
+        polycube-base:name-metric "router_ip_ports";
+        polycube-base:type-metric "G";
+        polycube-base:path-metric "$.ports[?(@.ip != '')]";
+        polycube-base:help-metric "Number of ports of the router with ip";
+        polycube-base:type-operation "FILTER";
       }
 
       list secondaryip {
         description "Additional IP addresses for the port";
         key "ip";
+        polycube-base:name-metric "router_secondaryip_ports";
+        polycube-base:type-metric "G";
+        polycube-base:path-metric '$.ports[0].secondaryip.length';
```

Comment:
Thanks, nice observation! (I made many changes over time and I was sure that something had escaped me. After the exams I will also check the other yangs)

This metric should be something like "number of secondary ip addresses of the router" with path equal to "$ .ports [*]. secondaryip.length" and also with operation type equal to "FILTER".

Maybe we could extend the metric as "number of secondary ip addresses of port X" but right now I know that it would take another label, so something like {cubeName = "r1", portsName = "p1"} = value. 

However, this would mean changing both the yang and the server because it is a double level of label. For now I would opt for the first solution, there will be changes in the future. I talk both at the level of adding metrics in the yang (in fact not all services have metrics) and at the level of creating metrics.

A "dratic"  choice that comes to me at the moment (I don't know how convenient) would be to make sure that there is all the information in the json of a cube to be able to create the metrics at run-time and then also add the value. This should also answer the doubt about the two functions on the server.

Let me know what you think and thanks a lot again for the observation @FedeParola :+1:

Reply:
> Thanks, nice observation! (I made many changes over time and I was sure that something had escaped me. After the exams I will also check the other yangs)
> 
> This metric should be something like "number of secondary ip addresses of the router" with path equal to "$ .ports [*]. secondaryip.length" and also with operation type equal to "FILTER".
 
I think this solution would be the best for now. Unfortunately JSONPath is not very flexible, and complex metrics require processing also on the server side (as the need of a `type-operation` field shows)

Reply:
OK thanks. I'll change that metric and update the PR.

About JSONPath you're absolutely right. At the beginning I was undecided on what to use, I found jsoncons and I got to work, only towards the end I started to understand its limitations. Maybe in the future it could be changed with some other library, like, I don't know, jq could be an alternative (It shouldn't be difficult to change the library, it's just a matter of changing the "path" in the metrics and something else)

Reply:
Hi @FedeParola . About my last osservation, I was wrong. I had some time to think about it and the correct solution is this:

```
polycube-base: path-metric 'sum ($. ports [*]. secondaryip.length)';
```
Without any "FILTER" as a type of operation. This is because, in this case I need the sum of the values ​​of the vector returned by "length".

Reply:
That's great, I didn't notice it but jsoncons extends the way aggregation functions can be used (the expression you wrote shouldn't work with the "standard" JSONPath). These aggregation functions should remove the limitations of JSONPath and it should be possible to completely drop the `type-operation` extension. For example here https://github.com/polycube-network/polycube/blob/8668f26b61b79678a0d64f5ea6a3b6ded008b518/src/services/pcn-router/datamodel/router.yang#L25-L30 we could just use the expression `count($.ports[?(@.ip != '')])` and remove the "FILTER", correct me if I'm wrong.

Reply:
I thought about it too, but the problem arises when filters and functions are put together. 

For example, it is not possible to apply "length" to a query that contains a filter and neither can ``` count($.ports[?(@.ip != '')]) ```. As far as I understand it is a lack of JSONPath. The way I have found to "fix" this problem is to use type-operation. Another possible way is to make two queries and in the second one give as input the output of the first query.

Reply:
You are right, functions don't support filter expressions, this is really unfortunate, it would have solved a lot of problems.